src/storage/lsm.rs

```diff
@@ -210,20 +217,74 @@ impl LSM {
     }
 
     fn compact_sstables(&mut self) -> Result<()> {
-        let mut dense_idx = Memtable::default();
-        for old_sst in self.sstables.iter() {
-            dense_idx.update_from_commit_log(&old_sst.path)?;
+
+        let path = new_path(SSTABLES_DIR_PATH);
+        let mut file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .open(&path)?;
+
+        // TODO(btc): for each old SSTable file, create iterator
+        // TODO(btc): while there exists an iterator with values, loop over them selecting the last instance of the smallest key.
+        // TODO(btc): write this key, value pair to the output file.
+
+        let mut iterators: Vec<Peekable<KeyValueIterator>> = Default::default();
+        for table in &self.sstables {
+            let p = &table.path;
+            let file = File::open(p)?;
+            let iter: KeyValueIterator = file.into();
+            let peekable = iter.peekable();
+            iterators.push(peekable);
         }
-        let new_sst = SSTable::write_from_memtable(&dense_idx, new_path(SSTABLES_DIR_PATH))?;
 
-        let old_sst_list: Vec<SSTable>;
-        {
-            // TODO MutexGuard here
-            // In async version, we will have to assume that new sstables may have been created while we were compacting, so we won't be able to just swap.
-            old_sst_list = mem::replace(&mut self.sstables, vec![new_sst]);
+        loop {
+            let mut min: Option<usize> = None;
+
+            for index in 0..iterators.len() {
+
+                if iterators[index].peek().is_none() { // iterator is depleted, so ignore it
+                    continue;
+                }
+
+                let (curr, _) = iterators[index].peek().unwrap();
+                let curr = curr.clone();
+
+                if min.is_none() { // then assign this key as min
+                    min = Some(index);
+                    continue;
+                }
+
+                // borrow
+                let min_key = iterators[min.unwrap()].peek().unwrap().0;
+
+                if curr <= min_key {
+                    // '<' less-than implies |curr| is smaller (the common case)
+                    // '==' equal implies |curr| is more recent (a subtle case)
+                    //
+                    // in both cases, update the min pointer and
+                    min = Some(index);
+                    continue;
+                }
+            }
+
+            if min.is_none() { // then all iterators are depleted
+                break; // out of the outermost loop, ending compaction
+            }
+
+            let index = min.unwrap();
+            if let Some((k, v)) = iterators[index].next() {
+                // TODO(btc): when is it valid to write (value = None) to `write_kv`?
```

Comment:
> when is it valid to write (value = None) to `write_kv`?

Before the recent refactor, (value == None) meant tombstone. We want to take the latest value whether it's a tombstone or not. So it's always safe to take the latest value. After the refactor, it's much more clear, as the write_kv (renamed seralize_kv) fn takes `&Value` and `Value::Tombstone` is a member of the enum.

Reply:
yay! makes sense!